prelude/cxx/tools/compile_commands.bxl

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under both the MIT license found in the
+# LICENSE-MIT file in the root directory of this source tree and the Apache
+# License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+# of this source tree.
+
+load("@prelude//:paths.bzl", "paths")
+load("@prelude//utils:set.bzl", "set")
+
+# Create a unified compilation database for all targets.  This is useful when trying to integrate
+# buck2 with traditional LSPs like clangd, which expect not a compilation database per target, but
+# rather a single global compilation database.  The user can optionally specify a target pattern
+# with --targets=<pattern> to generate a commpilation database for a subset of targets.
+#
+# This works by using aquery to query the action graph on each target in the selected target set,
+# and then filtering out actions with a category other than cxx_compile.
+#
+# It is possible for multiple targets to contain the same file.  In this case, we choose to select
+# only the first file and throw away any subsequent compilation database entries referring to the
+# same file.  LSP servers do not handle this case in general anyway.
+#
+# Presently the bxl writes the stdout, and the user is expected to redirect this to a file.  A
+# possible future improvement could perhaps provide an option to write it as an artifact to the
+# output folder and only print the path to the output file to stdout, so that it could benefit from
+# caching.
+#
+# The bxl is quite fast.  It has no dependencies on actually running any actions (unlike building
+# the [compilation-database] subtarget for a given target), and on a moderately sized project with
+# around 4000 source files, generation times on the order of 5 seconds or less can be expected.
+#
+# Sample Usage:
+#     buck2 bxl root//prelude/tools/bxl/compile_commands.bxl:gen_compilation_database > compile_commands.json
+#
+
+def _create_compilation_database_entry(ctx, buildfile_folder, action):
+
+    # the "cmd" attribute looks like a list.  For example, it might be something like:
+    #   cmd = [cl.exe, -c, foo.cpp]
+    # but in reality this is quite literally the string "[cl.exe, -c, foo.cpp]".  We need
+    # it as an actual list, so we have to employ this unfortunate hack of stripping the brackets,
+    # splitting on comma operator, and then stripping spaces from each entry to make an actual
+    # list.  Hopefully this doesn't fail in any weird edge cases, but it seems to be the best we
+    # can do until buck2 upstream actually stores a Starlark list as the value of this attribute.
+    cmd = action.attrs.cmd.value()
+    cmd = cmd.strip("[]").split(',')
+    cmd = list(map(lambda x: x.strip(), cmd))
```

Comment:
Can we get this chunk of code pulled out into a separate function? We probably will want to do something in future to clean this up, and having it separate makes it easier to see what to do.

CC @wendy728 for how we might want to unpack cmd_args values in future.